ThirdParty/netif/gmlc/netif/NetIF.hpp

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright © 2019,
+Battelle Memorial Institute; Lawrence Livermore National Security, LLC; Alliance
+for Sustainable Energy, LLC.  See the top-level NOTICE for additional details.
+All rights reserved. SPDX-License-Identifier: BSD-3-Clause
+*/
+
+#pragma once
+
+#include <string>
+#include <vector>
+#include <stdio.h>
+#include <stdlib.h>
+
+// other platforms might need different headers -- detecting with macros:
+// https://sourceforge.net/p/predef/wiki/OperatingSystems/
+// https://stackoverflow.com/questions/142508/how-do-i-check-os-with-a-preprocessor-directive
+#if defined(_WIN32)
+  #include <winsock2.h>
+  #include <WS2tcpip.h>
+  #include <iphlpapi.h>
+
+  #pragma comment(lib, "IPHLPAPI.lib")
+  #pragma comment(lib, "Ws2_32.lib")
+#else
+  #include <arpa/inet.h>
+  #include <ifaddrs.h>
+#endif
+
+namespace gmlc
+{
+namespace netif
+{
+  inline std::string addressToString(struct sockaddr *addr)
+  {
+    int family = addr->sa_family;
+    char addr_str[INET6_ADDRSTRLEN];
+    void *src_addr = nullptr;
+    switch (family) {
+      case AF_INET:
+        src_addr = &((struct sockaddr_in *)addr)->sin_addr;
+        break;
+      case AF_INET6:
+        src_addr = &((struct sockaddr_in6 *)addr)->sin6_addr;
+        break;
+      default: // Invalid address type for conversion to text
+        return std::string();
+    }
+    inet_ntop(family, src_addr, addr_str, INET6_ADDRSTRLEN);
+    return std::string(addr_str);
+  }
+	
+  template <class A>
```

Comment:
And the requirements on types for A?

Reply:
On Windows it is `PIP_ADAPTER_ADDRESSES` and on other platforms (macOS, Linux) it is `struct ifaddrs *`; I'm not sure if other platforms use something else for their interface address structure.

I'm not sure if using templates+decltype has any benefits or is just overcomplicating things -- do you think it would make sense to switch to using a typedef based on platform instead?

Reply:
That might be a bit clearer than a template with no restrictions?  I am sure it would error, but as a template I would have no idea how to use it.  with the typedefs ( or just completely separate functions) I could probably figure out by the arguments.